Agent configuration file and API restriction

sameo opened the pull request:
This PR adds 2 new features to the kata agent:

1. The ability to configure it through a configuration file from the root filesystem, referenced from the kernel command line (`agent.config_file`)
2. The ability to block part of the grpc API, through the configuration file.

Fixes #1837

egernst commented:
Can you highlight the design/reason for having the configuration inside the guest rootfs?

bergwolf commented on src/agent/src/main.rs:
```diff
@@ -77,11 +77,11 @@ mod rpc;
-    static ref AGENT_CONFIG: Arc<RwLock<AgentConfig>> =
-        Arc::new(RwLock::new(config::AgentConfig::new()));
+    static ref AGENT_CONFIG: Arc<RwLock<AgentConfig>> = Arc::new(RwLock::new(
+        AgentConfig::from_cmdline("/proc/cmdline").unwrap()
```
Can you assign it later and bail out properly when failing to parse the config file?

sameo replied:
@bergwolf Assignment is now done later. It was causing the initial PR to fail as we were trying to parse cmdline before /proc was mounted.

bergwolf commented on src/agent/src/config.rs:
```diff
@@ -59,6 +67,76 @@ pub struct AgentConfig {
+}
+
+#[derive(Debug, Deserialize)]
+pub struct AgentConfigBuilder {
```
IMO, it should have a sane default for some of the fields, like `hotplug_timeout`, `debug_console_vport`, `log_vport` and `server_addr`.

sameo replied:
@bergwolf The `AgentConfigBuilder` will only overwrites what's coming from a default `AgentConfig`. In other words, when building an `AgentConfig` from a configuration file, we start from a defaulted `AgentConfig` and the configuration file overwrites the default values. It's like an overlay.

bergwolf reviewed (commented):
Overall lgtm. Just two minor comments.

GabyCT commented:
/test-ubuntu-pmem

sameo commented:
> Can you highlight the design/reason for having the configuration inside the guest rootfs?

Sorry, I forgot to add the issue it fixes, see #1837. This is for implementing the API restriction based on a measured configuration (i.e. coming from the initrd/rootfs).

sameo commented:
/test

sameo commented:
/test

sameo commented:
/test

sameo commented:
/test

sameo commented:
/test-clh-k8s-containerd

sameo commented:
@bergwolf Could you PTAL?

dcmiddle commented:
Great to see this item from the CCv0 plan so far along.
Unfortunately, I'm pretty late to this issue and PR, but have added a couple comments to the issue. Namely that an allow list / default-deny policy is generally more secure. 
An alternative to stay with the deny-list approach currently in this PR would be to provide a restrictive default example.

Also, regarding an example, it would be nice to have a `agent-configuration.toml` as part of this PR, even if its contents were still open to debate. I see there's a separate issue created to discuss the security implications of the APIs (https://github.com/kata-containers/kata-containers/issues/1891).

Jakob-Naucke reviewed (commented):
LGTM although we should have more than two pairs of eyes for this I think :)

mxpv commented on src/agent/src/config.rs:
```diff
@@ -59,6 +68,80 @@ pub struct AgentConfig {
+
+impl AgentConfigBuilder {
+    #[instrument]
+    fn from_config_string(config: &str) -> Result<AgentConfig> {
```
`FromStr` trait would be nicer here? https://doc.rust-lang.org/std/str/trait.FromStr.html

sameo commented on src/agent/src/config.rs:
```diff
@@ -59,6 +67,76 @@ pub struct AgentConfig {
+        );
+        config_override!(agent_config_builder, agent_config, hotplug_timeout);
+        config_override!(agent_config_builder, agent_config, debug_console_vport);
+        config_override!(agent_config_builder, agent_config, log_vport);
```
I'd have to look more into that, as I'm passing identifier (field names) as the last argument. I may be able to come up with another macro that'd export the `AgentConfig` fields, but this could be fairly complex. More so as this PR is not treating all fields the same way...

Jakob-Naucke replied:
No, I see. Alright.

sameo commented:
> Great to see this item from the CCv0 plan so far along. Unfortunately, I'm pretty late to this issue and PR, but have added a couple comments to the issue. Namely that an allow list / default-deny policy is generally more secure. An alternative to stay with the deny-list approach currently in this PR would be to provide a restrictive default example.
> 

You're right, and @magowan was also in favor of an allowed list rather than a blocked one.
So I changed the PR so that if there is an `[endpoints] section in the configuration file, then we add the allowed endpoints to the hash set. If there is no such section, we assume all endpoints are not allowed.


> Also, regarding an example, it would be nice to have a `agent-configuration.toml` as part of this PR, even if its contents were still open to debate. I see there's a separate issue created to discuss the security implications of the APIs (#1891).

I also added a example of a configuration file, with all endpoints allowed. Then we can add a reference one as an outcome of the #1891 discussions.

sameo commented:
/test

sameo commented:
/test

Jakob-Naucke commented:
> The s390x and ARM CI are known to be broken and pending a few static check fixes. @egernst @Jakob-Naucke could you PTAL?

@sameo #2774

sameo commented:
/test
